packages/ckb-sdk-core/src/index.ts

```diff
@@ -218,8 +262,201 @@ class Core {
+    }
+    if (!this.config.daoDep) {
+      await this.loadDaoDep()
+    }
+
+    const DAO_LOCK_PERIOD_EPOCHS = 180
+    const cellStatus = await this.rpc.getLiveCell(withdrawOutPoint, true)
+    if (cellStatus.status !== 'live') throw new Error('Cell is not live yet')
+
+    const tx = await this.rpc.getTransaction(withdrawOutPoint.txHash)
+    if (tx.txStatus.status !== 'committed') throw new Error('Transaction is not committed yet')
+
+    /* eslint-disable */
+    const depositBlockNumber = +this.utils.bytesToHex(this.utils.hexToBytes(cellStatus.cell.data?.content ?? '').reverse())
+    /* eslint-enable */
+
+    const depositBlockHeader = await this.rpc.getBlockByNumber(BigInt(depositBlockNumber)).then(block => block.header)
+    const depositEpoch = this.utils.parseEpoch(depositBlockHeader.epoch)
+
+    const withdrawBlockHeader = await this.rpc.getBlock(tx.txStatus.blockHash).then(block => block.header)
+    const withdrawEpoch = this.utils.parseEpoch(withdrawBlockHeader.epoch)
+
+    const withdrawFraction = withdrawEpoch.index * withdrawEpoch.length
+    const depositFraction = depositEpoch.index * depositEpoch.length
+    let depositedEpochs = withdrawEpoch.number - depositEpoch.number
+    if (withdrawFraction > depositFraction) {
+      depositedEpochs += 1
+    }
+
+    const lockEpochs = Math.floor((depositedEpochs + DAO_LOCK_PERIOD_EPOCHS - 1) / 10) * 10
+    const minimalSinceEpochNumber = depositEpoch.number + lockEpochs
+    const minimalSinceEpochIndex = depositEpoch.index
+    const minimalSinceEpochLength = depositEpoch.length
+
+    const minimalSince = this.epochSince(minimalSinceEpochLength, minimalSinceEpochIndex, minimalSinceEpochNumber)
+    const outputCapacity = await this.rpc.calculateDaoMaximumWithdraw(depositOutPoint, withdrawBlockHeader.hash)
+    const targetCapacity = BigInt(outputCapacity)
+    const targetFee = BigInt(fee)
+    if (targetCapacity < targetFee) {
+      throw new Error(`The fee(${targetFee}) is too big that withdraw(${targetCapacity}) is not enough`)
+    }
+
+    const outputs: CKBComponents.CellOutput[] = [
+      {
+        capacity: `0x${(targetCapacity - targetFee).toString(16)}`,
+        lock: tx.transaction.outputs[+withdrawOutPoint.index].lock,
+      },
+    ]
+
+    const outputsData = ['0x']
+
+    return {
+      version: '0x0',
+      cellDeps: [
+        { outPoint: this.config.secp256k1Dep!.outPoint, depType: 'depGroup' },
+        { outPoint: this.config.daoDep!.outPoint, depType: 'code' },
+      ],
+      headerDeps: [
+        depositBlockHeader.hash,
+        withdrawBlockHeader.hash,
+      ],
+      inputs: [
+        {
+          previousOutput: withdrawOutPoint,
+          since: minimalSince,
+        },
+      ],
+      outputs,
+      outputsData,
+      witnesses: [{
+        lock: '',
+        inputType: '0x0000000000000000',
+        outputType: '',
+      }],
+    }
+  }
+
+  private epochSince = (length: number, index: number, number: number) =>
```

Comment:
Suggest renaming to `absoluteEpochSince`, to distinguish relative-epoch-since.

Reply:
updated in https://github.com/nervosnetwork/ckb-sdk-js/pull/382/commits/5580419c387f5202d71045a027ff7505d3bbdd09